#### source : https://www.kaggle.com/datasets/heesoo37/120-years-of-olympic-history-athletes-and-results

# 1) This is a historical dataset on the modern Olympic Games, including all the Games from Athens 1896 to Rio 2016. It contains 271116 rows and 15 columns. 
### Note : the Winter and Summer Games were held in the same year up until 1992. After that, they staggered them such that Winter Games occur on a four year cycle starting with 1994, then Summer in 1996, then Winter in 1998, and so on. A common mistake people make when analyzing this data is to assume that the Summer and Winter Games have always been staggered.

## We will analyze this dataset by querying it using SQL. We will cover up to 20 queries.

## 1.1) Before performing any query, we need to set up our environment. It is indeed possible to set up an SQL environment using Jupyter Notebook, since Python provides an extension for it. One way to do this is by using the ipython-sql library.

In [1]:
%load_ext sql

In [2]:
import pandas as pd
athlete_events = pd.read_csv("/home/loick/Downloads/archive/athlete_events.csv")
athlete_events.head(10)

ID                      Name Sex   Age  Height  Weight            Team  \
0   1                 A Dijiang   M  24.0   180.0    80.0           China   
1   2                  A Lamusi   M  23.0   170.0    60.0           China   
2   3       Gunnar Nielsen Aaby   M  24.0     NaN     NaN         Denmark   
3   4      Edgar Lindenau Aabye   M  34.0     NaN     NaN  Denmark/Sweden   
4   5  Christine Jacoba Aaftink   F  21.0   185.0    82.0     Netherlands   
5   5  Christine Jacoba Aaftink   F  21.0   185.0    82.0     Netherlands   
6   5  Christine Jacoba Aaftink   F  25.0   185.0    82.0     Netherlands   
7   5  Christine Jacoba Aaftink   F  25.0   185.0    82.0     Netherlands   
8   5  Christine Jacoba Aaftink   F  27.0   185.0    82.0     Netherlands   
9   5  Christine Jacoba Aaftink   F  27.0   185.0    82.0     Netherlands   

   NOC        Games  Year  Season         City          Sport  \
0  CHN  1992 Summer  1992  Summer    Barcelona     Basketball   
1  CHN  2012 Summer  2012  Summer       London           Judo   
2  DEN  1920 Summer  1920  Summer    Antwerpen       Football   
3  DEN  1900 Summer  1900  Summer        Paris     Tug-Of-War   
4  NED  1988 Winter  1988  Winter      Calgary  Speed Skating   
5  NED  1988 Winter  1988  Winter      Calgary  Speed Skating   
6  NED  1992 Winter  1992  Winter  Albertville  Speed Skating   
7  NED  1992 Winter  1992  Winter  Albertville  Speed Skating   
8  NED  1994 Winter  1994  Winter  Lillehammer  Speed Skating   
9  NED  1994 Winter  1994  Winter  Lillehammer  Speed Skating   

                                Event Medal  
0         Basketball Men's Basketball   NaN  
1        Judo Men's Extra-Lightweight   NaN  
2             Football Men's Football   NaN  
3         Tug-Of-War Men's Tug-Of-War  Gold  
4    Speed Skating Women's 500 metres   NaN  
5  Speed Skating Women's 1,000 metres   NaN  
6    Speed Skating Women's 500 metres   NaN  
7  Speed Skating Women's 1,000 metres   NaN  
8    Speed Skating Women's 500 metres   NaN  
9  Speed Skating Women's 1,000 metres   NaN

## 1.2) Below, we go on by connecting to a database. Thus, we fill the data from our dataframe into a table inside this database that we created.

In [3]:
import sqlite3

# Connect to the database
conn = sqlite3.connect("athlete_events.db")

# Write the data from the dataframe to a table in the database
athlete_events.to_sql("athlete_events", con=conn, if_exists="replace")

# Close the connection
conn.close()

## 1.3) We can now connect to the database

In [4]:
%sql sqlite:///athlete_events.db

# 2) Our environement is ready. But before being able to query anything, we need to point out one specificity that is noticeable while reading through the table. 

## In this table are the countries that participated in olympic games, but also teams inside each country. For example : " Austria - 1 ", " Austria - 2 ". If we count the 'Teams' column that participated, we will potentially count each nation several times.

## That is why we will count the " NOC " column, that is similar either for a country or for one of its teams. Then we can apply the 'DISTINCT' statement to count the unique occurrences of each nation.

# That being said, we can start querying.

## In SQL, it is a necessity to phrase the query before coding it. If the phrase is transformed into a clear sentence, generally the query becomes self-evident to be coded.


## 2.1) How many olympics games have been held?

### For our case, let's reason logcially. We want to count how many olympic games have been held. We exactly have a column called "Games" that shows the year and the season of the game (Summer, Winter). 

### It is then enough to solve this query to just count the unique occurrence of each row in the "Games" column.

In [15]:
query_1 = """ 
SELECT COUNT(DISTINCT Games) AS 'Number of olypmic games'
FROM athlete_events;
"""

%sql $query_1

 * sqlite:///athlete_events.db
Done.


[(51,)]

## 2.2) We want to list down all Olympics games held so far

### Let's rephrase the query : we want to output a chart containing, for each row, the UNIQUE combination of {Year; season; city} of each game held

In [16]:
query_2 = """
SELECT DISTINCT Year, Season, City
FROM athlete_events
ORDER BY Year;
"""
%sql $query_2

 * sqlite:///athlete_events.db
Done.


[(1896, 'Summer', 'Athina'),
 (1900, 'Summer', 'Paris'),
 (1904, 'Summer', 'St. Louis'),
 (1906, 'Summer', 'Athina'),
 (1908, 'Summer', 'London'),
 (1912, 'Summer', 'Stockholm'),
 (1920, 'Summer', 'Antwerpen'),
 (1924, 'Summer', 'Paris'),
 (1924, 'Winter', 'Chamonix'),
 (1928, 'Summer', 'Amsterdam'),
 (1928, 'Winter', 'Sankt Moritz'),
 (1932, 'Summer', 'Los Angeles'),
 (1932, 'Winter', 'Lake Placid'),
 (1936, 'Summer', 'Berlin'),
 (1936, 'Winter', 'Garmisch-Partenkirchen'),
 (1948, 'Summer', 'London'),
 (1948, 'Winter', 'Sankt Moritz'),
 (1952, 'Summer', 'Helsinki'),
 (1952, 'Winter', 'Oslo'),
 (1956, 'Winter', "Cortina d'Ampezzo"),
 (1956, 'Summer', 'Melbourne'),
 (1956, 'Summer', 'Stockholm'),
 (1960, 'Winter', 'Squaw Valley'),
 (1960, 'Summer', 'Roma'),
 (1964, 'Winter', 'Innsbruck'),
 (1964, 'Summer', 'Tokyo'),
 (1968, 'Summer', 'Mexico City'),
 (1968, 'Winter', 'Grenoble'),
 (1972, 'Summer', 'Munich'),
 (1972, 'Winter', 'Sapporo'),
 (1976, 'Summer', 'Montreal'),
 (1976, 'Winter', 'Innsbruck'),
 (1980, 'Winter', 'Lake Placid'),
 (1980, 'Summer', 'Moskva'),
 (1984, 'Winter', 'Sarajevo'),
 (1984, 'Summer', 'Los Angeles'),
 (1988, 'Winter', 'Calgary'),
 (1988, 'Summer', 'Seoul'),
 (1992, 'Summer', 'Barcelona'),
 (1992, 'Winter', 'Albertville'),
 (1994, 'Winter', 'Lillehammer'),
 (1996, 'Summer', 'Atlanta'),
 (1998, 'Winter', 'Nagano'),
 (2000, 'Summer', 'Sydney'),
 (2002, 'Winter', 'Salt Lake City'),
 (2004, 'Summer', 'Athina'),
 (2006, 'Winter', 'Torino'),
 (2008, 'Summer', 'Beijing'),
 (2010, 'Winter', 'Vancouver'),
 (2012, 'Summer', 'London'),
 (2014, 'Winter', 'Sochi'),
 (2016, 'Summer', 'Rio de Janeiro')]

## 2.3) What is the total number of nations who participated in each olympics game?

### Let's rephrase. In each olympic game, several nations participate. We want a chart that shows, for each olympic game, the number of nations that participated. Translated into SQL, we need to COUNT each unique occurrence for a nation GROUPED BY olympic game.

In [17]:
query_3 = """
SELECT Games, COUNT(DISTINCT NOC) AS 'Number of nations' 
FROM athlete_events 
GROUP BY Games;
"""

%sql $query_3

 * sqlite:///athlete_events.db
Done.


[('1896 Summer', 12),
 ('1900 Summer', 31),
 ('1904 Summer', 15),
 ('1906 Summer', 21),
 ('1908 Summer', 22),
 ('1912 Summer', 29),
 ('1920 Summer', 29),
 ('1924 Summer', 45),
 ('1924 Winter', 19),
 ('1928 Summer', 46),
 ('1928 Winter', 25),
 ('1932 Summer', 47),
 ('1932 Winter', 17),
 ('1936 Summer', 49),
 ('1936 Winter', 28),
 ('1948 Summer', 59),
 ('1948 Winter', 28),
 ('1952 Summer', 69),
 ('1952 Winter', 30),
 ('1956 Summer', 72),
 ('1956 Winter', 32),
 ('1960 Summer', 84),
 ('1960 Winter', 30),
 ('1964 Summer', 93),
 ('1964 Winter', 36),
 ('1968 Summer', 112),
 ('1968 Winter', 37),
 ('1972 Summer', 121),
 ('1972 Winter', 35),
 ('1976 Summer', 92),
 ('1976 Winter', 37),
 ('1980 Summer', 80),
 ('1980 Winter', 37),
 ('1984 Summer', 140),
 ('1984 Winter', 49),
 ('1988 Summer', 159),
 ('1988 Winter', 57),
 ('1992 Summer', 169),
 ('1992 Winter', 64),
 ('1994 Winter', 67),
 ('1996 Summer', 197),
 ('1998 Winter', 72),
 ('2000 Summer', 200),
 ('2002 Winter', 77),
 ('2004 Summer', 201),
 ('2006 Winter', 79),
 ('2008 Summer', 204),
 ('2010 Winter', 82),
 ('2012 Summer', 205),
 ('2014 Winter', 89),
 ('2016 Summer', 207)]

### To check this result, let's count the total number of teams that compete for one game, for example the 1900 Summer Games. If what we find is the same number as above, we wrote our query correctly. Otherwise, something is wrong.

In [18]:
query3 = """
SELECT COUNT(*) AS 'Number of teams that participated in the 1900 Summer'
FROM (
  SELECT DISTINCT NOC, Games
  FROM athlete_events
  WHERE Games = '1900 Summer'
);
"""
%sql $query3

 * sqlite:///athlete_events.db
Done.


[(31,)]

### We find 31 again. Our query was correct.


## 2.4) Which year saw the highest no of countries participating in olympics ?

In [19]:
query_4 = """
SELECT Games, COUNT(DISTINCT NOC) AS 'Number of countries'
FROM athlete_events
GROUP BY Games
ORDER BY COUNT(DISTINCT Team) DESC
LIMIT 1
"""

%sql $query_4

 * sqlite:///athlete_events.db
Done.


[('2008 Summer', 204)]

##  What about the year that saw the lowest no of countries participating in olympics ?

In [20]:
query_4_1 = """
SELECT Games, COUNT(DISTINCT NOC) AS 'Number of countries'
FROM athlete_events
GROUP BY Games
ORDER BY COUNT(DISTINCT Team) ASC
LIMIT 1
"""

%sql $query_4_1

 * sqlite:///athlete_events.db
Done.


[('1896 Summer', 12)]

## 2.5) Which nation has participated in all of the olympic games ?

### rephrased : We look to know which nation attended each olympic game. since the oldest we have in our dataset.

## We know from 2.1) that a nation that attended all olympic games is the nation that attended exactly 51 olympic games.

In [21]:
query_5 = """
SELECT Team, COUNT(DISTINCT Games) AS 'Number of olympic participation'
FROM athlete_events
GROUP BY Team
HAVING COUNT(DISTINCT Games) = (SELECT COUNT(DISTINCT Games) FROM athlete_events)
"""

%sql $query_5

 * sqlite:///athlete_events.db
Done.


[('France', 51), ('Great Britain', 51), ('Italy', 51), ('Switzerland', 51)]

## 2.6) Which sport has been played in all summer olympics ?

### rephrased : We want to find the sport that has existed in each 'summer' game.

### Let's first display how many times we have had each sport among all the olympic games.

In [22]:
query_6 = """
SELECT Sport, SUM(NumOccurrences) AS TotalOccurrences
FROM (
  SELECT Sport, Games, COUNT(DISTINCT Sport) AS NumOccurrences
  FROM athlete_events
  GROUP BY Sport, Games
)
GROUP BY Sport
ORDER BY TotalOccurrences DESC
"""

%sql $query_6

 * sqlite:///athlete_events.db
Done.


[('Swimming', 29),
 ('Gymnastics', 29),
 ('Fencing', 29),
 ('Cycling', 29),
 ('Athletics', 29),
 ('Wrestling', 28),
 ('Rowing', 28),
 ('Water Polo', 27),
 ('Shooting', 27),
 ('Football', 27),
 ('Diving', 27),
 ('Weightlifting', 26),
 ('Sailing', 26),
 ('Equestrianism', 25),
 ('Boxing', 25),
 ('Modern Pentathlon', 24),
 ('Figure Skating', 24),
 ('Ice Hockey', 23),
 ('Hockey', 23),
 ('Speed Skating', 22),
 ('Ski Jumping', 22),
 ('Nordic Combined', 22),
 ('Cross Country Skiing', 22),
 ('Bobsleigh', 21),
 ('Canoeing', 19),
 ('Basketball', 19),
 ('Alpine Skiing', 19),
 ('Tennis', 16),
 ('Archery', 16),
 ('Biathlon', 15),
 ('Volleyball', 14),
 ('Luge', 14),
 ('Judo', 13),
 ('Handball', 13),
 ('Synchronized Swimming', 9),
 ('Rhythmic Gymnastics', 9),
 ('Table Tennis', 8),
 ('Short Track Speed Skating', 7),
 ('Freestyle Skiing', 7),
 ('Badminton', 7),
 ('Art Competitions', 7),
 ('Tug-Of-War', 6),
 ('Skeleton', 6),
 ('Curling', 6),
 ('Beach Volleyball', 6),
 ('Triathlon', 5),
 ('Trampolining', 5),
 ('Taekwondo', 5),
 ('Snowboarding', 5),
 ('Polo', 5),
 ('Baseball', 5),
 ('Softball', 4),
 ('Rugby', 4),
 ('Golf', 3),
 ('Alpinism', 3),
 ('Lacrosse', 2),
 ('Rugby Sevens', 1),
 ('Roque', 1),
 ('Racquets', 1),
 ('Motorboating', 1),
 ('Military Ski Patrol', 1),
 ('Jeu De Paume', 1),
 ('Croquet', 1),
 ('Cricket', 1),
 ('Basque Pelota', 1),
 ('Aeronautics', 1)]

## The answer to our query are the maximum amounts that we can find from this query above. 

## 2.7) Which Sports were just played only once in the olympics ?

In [23]:
query_7 = """
SELECT Sport, COUNT(Sport), Games
FROM (
  SELECT Sport, Games, COUNT(DISTINCT Sport) AS NumOccurrences
  FROM athlete_events
  GROUP BY Sport, Games
)
GROUP BY Sport
HAVING COUNT(NumOccurrences) = 1
"""

%sql $query_7

 * sqlite:///athlete_events.db
Done.


[('Aeronautics', 1, '1936 Summer'),
 ('Basque Pelota', 1, '1900 Summer'),
 ('Cricket', 1, '1900 Summer'),
 ('Croquet', 1, '1900 Summer'),
 ('Jeu De Paume', 1, '1908 Summer'),
 ('Military Ski Patrol', 1, '1924 Winter'),
 ('Motorboating', 1, '1908 Summer'),
 ('Racquets', 1, '1908 Summer'),
 ('Roque', 1, '1904 Summer'),
 ('Rugby Sevens', 1, '2016 Summer')]

## 2.8) Let's fetch the total number of sports played in each olympic games

In [24]:
query_8 = """
SELECT Games, COUNT(DISTINCT Sport) AS TotalSports
FROM athlete_events
GROUP BY Games
ORDER BY TotalSports DESC
"""

%sql $query_8

 * sqlite:///athlete_events.db
Done.


[('2016 Summer', 34),
 ('2008 Summer', 34),
 ('2004 Summer', 34),
 ('2000 Summer', 34),
 ('2012 Summer', 32),
 ('1996 Summer', 31),
 ('1992 Summer', 29),
 ('1988 Summer', 27),
 ('1984 Summer', 25),
 ('1920 Summer', 25),
 ('1936 Summer', 24),
 ('1908 Summer', 24),
 ('1980 Summer', 23),
 ('1976 Summer', 23),
 ('1972 Summer', 23),
 ('1964 Summer', 21),
 ('1968 Summer', 20),
 ('1948 Summer', 20),
 ('1924 Summer', 20),
 ('1900 Summer', 20),
 ('1960 Summer', 19),
 ('1956 Summer', 19),
 ('1952 Summer', 19),
 ('1932 Summer', 18),
 ('1904 Summer', 18),
 ('1928 Summer', 17),
 ('1912 Summer', 17),
 ('2014 Winter', 15),
 ('2010 Winter', 15),
 ('2006 Winter', 15),
 ('2002 Winter', 15),
 ('1998 Winter', 14),
 ('1906 Summer', 13),
 ('1994 Winter', 12),
 ('1992 Winter', 12),
 ('1988 Winter', 10),
 ('1984 Winter', 10),
 ('1980 Winter', 10),
 ('1976 Winter', 10),
 ('1972 Winter', 10),
 ('1968 Winter', 10),
 ('1964 Winter', 10),
 ('1924 Winter', 10),
 ('1948 Winter', 9),
 ('1896 Summer', 9),
 ('1960 Winter', 8),
 ('1956 Winter', 8),
 ('1952 Winter', 8),
 ('1936 Winter', 8),
 ('1928 Winter', 8),
 ('1932 Winter', 7)]

## 2.9) Let's fetch the oldest athletes to win a gold medal

In [25]:
query_9 = """
SELECT Name, Age, Year
FROM athlete_events
WHERE Medal = 'Gold' AND Age = (SELECT MAX(Age) FROM athlete_events WHERE Medal = 'Gold' AND Year = athlete_events.Year)
GROUP BY Year
"""

%sql $query_9

 * sqlite:///athlete_events.db
Done.


[('Charles Jacobus', 64.0, 1904), ('Oscar Gomer Swahn', 64.0, 1912)]

## 2.10) Let's find the ratio of male and female athletes who participated in all olympic games.

In [26]:
query_10 = """
SELECT
  CAST(COUNT(CASE WHEN Sex = 'M' THEN 1 END) AS REAL) AS NumMales,
  CAST(COUNT(CASE WHEN Sex = 'F' THEN 1 END) AS REAL) AS NumFemales,
  ROUND(CAST(COUNT(CASE WHEN Sex = 'M' THEN 1 END) AS REAL) / NULLIF(CAST(COUNT(CASE WHEN Sex = 'F' THEN 1 END) AS REAL), 0), 3) AS MaleToFemaleRatio
FROM athlete_events
"""

%sql $query_10

 * sqlite:///athlete_events.db
Done.


[(196594.0, 74522.0, 2.638)]

## There are approximately 2.638 more men than women who participated in olympic games according to our dataframe

## 2.11) Let's fetch the top 5 athletes who have won the most gold medals.

In [27]:
query_11 = """
SELECT Name, NumGoldMedals, Ranking
FROM (
  SELECT Name, COUNT(*) AS NumGoldMedals,
  DENSE_RANK() OVER (ORDER BY COUNT(*) DESC) AS Ranking
  FROM athlete_events
  WHERE Medal = 'Gold'
  GROUP BY Name
) AS ranked_athletes
WHERE Ranking <= 5
"""

%sql $query_11

 * sqlite:///athlete_events.db
Done.


[('Michael Fred Phelps, II', 23, 1),
 ('Raymond Clarence "Ray" Ewry', 10, 2),
 ('Paavo Johannes Nurmi', 9, 3),
 ('Mark Andrew Spitz', 9, 3),
 ('Larysa Semenivna Latynina (Diriy-)', 9, 3),
 ('Frederick Carlton "Carl" Lewis', 9, 3),
 ('Usain St. Leo Bolt', 8, 4),
 ('Sawao Kato', 8, 4),
 ('Ole Einar Bjrndalen', 8, 4),
 ('Matthew Nicholas "Matt" Biondi', 8, 4),
 ('Jennifer Elisabeth "Jenny" Thompson (-Cumpelik)', 8, 4),
 ('Birgit Fischer-Schmidt', 8, 4),
 ('Vra slavsk (-Odloilov)', 7, 5),
 ('Viktor Ivanovych Chukarin', 7, 5),
 ('Nikolay Yefimovich Andrianov', 7, 5),
 ('Donald Arthur "Don" Schollander', 7, 5),
 ('Borys Anfiyanovych Shakhlin', 7, 5),
 ('Aladr Gerevich (-Gerei)', 7, 5)]

## We notice that there is only one number one, but several other numbers 2 or 3 ;)

## 2.12) Let's fetch the top 5 athletes who have won the most medals (gold/silver/bronze)

### let's rephrase : we are asked to count the number of medals of any kind for each athlete and set a ranking out of this request. We just need to adapt the code above for any medal.

In [28]:
query_12 = """
SELECT Name, Team, NumMedals, Ranking
FROM (
  SELECT Name, Team, COUNT(*) AS NumMedals,
  DENSE_RANK() OVER (ORDER BY COUNT(*) DESC) AS Ranking
  FROM athlete_events
  WHERE Medal IN ('Gold', 'Silver', 'Bronze')
  GROUP BY Name
) AS ranked_athletes
WHERE Ranking <= 5
"""

%sql $query_12

 * sqlite:///athlete_events.db
Done.


[('Michael Fred Phelps, II', 'United States', 28, 1),
 ('Larysa Semenivna Latynina (Diriy-)', 'Soviet Union', 18, 2),
 ('Nikolay Yefimovich Andrianov', 'Soviet Union', 15, 3),
 ('Takashi Ono', 'Japan', 13, 4),
 ('Ole Einar Bjrndalen', 'Norway', 13, 4),
 ('Edoardo Mangiarotti', 'Italy', 13, 4),
 ('Borys Anfiyanovych Shakhlin', 'Soviet Union', 13, 4),
 ('Sawao Kato', 'Japan', 12, 5),
 ('Ryan Steven Lochte', 'United States', 12, 5),
 ('Paavo Johannes Nurmi', 'Finland', 12, 5),
 ('Natalie Anne Coughlin (-Hall)', 'United States', 12, 5),
 ('Jennifer Elisabeth "Jenny" Thompson (-Cumpelik)', 'United States', 12, 5),
 ('Dara Grace Torres (-Hoffman, -Minas)', 'United States', 12, 5),
 ('Birgit Fischer-Schmidt', 'East Germany', 12, 5),
 ('Aleksey Yuryevich Nemov', 'Russia', 12, 5)]

## 2.13) Now let's fetch the top 5 most successful countries in olympics. Success is defined by no of medals won

In [29]:
query_13 = """
SELECT Team, NumMedals, Ranking
FROM (
  SELECT
    (CASE WHEN Team IN ('Soviet Union','Russia') THEN 'Russia/Soviet Union' ELSE Team END) AS Country,
    Team, COUNT(*) AS NumMedals,
    DENSE_RANK() OVER (ORDER BY COUNT(*) DESC) AS Ranking
  FROM athlete_events
  WHERE Medal IN ('Gold', 'Silver', 'Bronze')
  GROUP BY Team
) AS ranked_countries
WHERE Ranking <= 5
"""

%sql $query_13

 * sqlite:///athlete_events.db
Done.


[('United States', 5219, 1),
 ('Soviet Union', 2451, 2),
 ('Germany', 1984, 3),
 ('Great Britain', 1673, 4),
 ('France', 1550, 5)]

## 2.14) Let's list down total gold, silver and bronze medals won by each country

In [30]:
query_14 = """
SELECT NOC, Team,
  SUM(CASE WHEN Medal = 'Gold' THEN 1 ELSE 0 END) AS NumGoldMedals,
  SUM(CASE WHEN Medal = 'Silver' THEN 1 ELSE 0 END) AS NumSilverMedals,
  SUM(CASE WHEN Medal = 'Bronze' THEN 1 ELSE 0 END) AS NumBronzeMedals,
  DENSE_RANK() OVER (ORDER BY SUM(CASE WHEN Medal IN ('Gold', 'Silver', 'Bronze') THEN 1 ELSE 0 END) DESC) AS Ranking
FROM athlete_events
GROUP BY NOC
"""

%sql $query_14

 * sqlite:///athlete_events.db
Done.


[('USA', 'United States', 2638, 1641, 1358, 1),
 ('URS', 'Soviet Union', 1082, 732, 689, 2),
 ('GER', 'Germany', 745, 674, 746, 3),
 ('GBR', 'Great Britain', 678, 739, 651, 4),
 ('FRA', 'France', 501, 610, 666, 5),
 ('ITA', 'Italy', 575, 531, 531, 6),
 ('SWE', 'Sweden', 479, 522, 535, 7),
 ('CAN', 'Canada', 463, 438, 451, 8),
 ('AUS', 'Australia', 348, 455, 517, 9),
 ('RUS', 'Russia', 390, 367, 408, 10),
 ('HUN', 'Hungary', 432, 332, 371, 11),
 ('NED', 'Netherlands', 287, 340, 413, 12),
 ('NOR', 'Norway', 378, 361, 294, 13),
 ('GDR', 'East Germany', 397, 327, 281, 14),
 ('CHN', 'China', 350, 347, 292, 15),
 ('JPN', 'Japan', 247, 309, 357, 16),
 ('FIN', 'Finland', 198, 270, 432, 17),
 ('SUI', 'Switzerland', 175, 248, 268, 18),
 ('ROU', 'Romania', 161, 200, 292, 19),
 ('KOR', 'South Korea', 221, 232, 185, 20),
 ('DEN', 'Denmark', 179, 241, 177, 21),
 ('FRG', 'West Germany', 159, 194, 233, 22),
 ('POL', 'Poland', 117, 195, 253, 23),
 ('ESP', 'Spain', 110, 243, 136, 24),
 ('TCH', 'Czechoslovakia', 81, 225, 182, 25),
 ('BRA', 'Brazil', 109, 175, 191, 26),
 ('BEL', 'Belgium', 98, 197, 173, 27),
 ('AUT', 'Austria', 108, 186, 156, 28),
 ('CUB', 'Cuba', 164, 129, 116, 29),
 ('YUG', 'Yugoslavia', 130, 167, 93, 30),
 ('BUL', 'Bulgaria', 54, 144, 144, 31),
 ('EUN', 'Unified Team', 127, 71, 81, 32),
 ('ARG', 'Argentina', 91, 92, 91, 33),
 ('GRE', 'Greece', 62, 109, 84, 34),
 ('NZL', 'New Zealand', 90, 56, 82, 35),
 ('UKR', 'Ukraine', 47, 52, 100, 36),
 ('IND', 'India', 138, 19, 40, 37),
 ('JAM', 'Jamaica', 38, 75, 44, 38),
 ('CRO', 'Croatia', 58, 54, 37, 39),
 ('CZE', 'Czech Republic', 42, 36, 66, 40),
 ('BLR', 'Belarus', 24, 44, 71, 41),
 ('RSA', 'South Africa', 32, 47, 52, 42),
 ('PAK', 'Pakistan', 42, 45, 34, 43),
 ('MEX', 'Mexico', 30, 26, 54, 44),
 ('KEN', 'Kenya', 34, 41, 31, 45),
 ('NGR', 'Nigeria', 23, 30, 46, 46),
 ('TUR', 'Turkey', 40, 27, 28, 47),
 ('SRB', 'Serbia', 15, 29, 41, 48),
 ('KAZ', 'Kazakhstan', 20, 25, 32, 49),
 ('IRI', 'Iran', 18, 21, 29, 50),
 ('PRK', 'North Korea', 16, 16, 35, 51),
 ('SCG', 'Serbia and Montenegro', 12, 26, 26, 52),
 ('URU', 'Uruguay', 31, 2, 30, 53),
 ('LTU', 'Lithuania', 6, 7, 48, 54),
 ('ETH', 'Ethiopia', 22, 9, 22, 55),
 ('EST', 'Estonia', 13, 12, 25, 56),
 ('TPE', 'Chinese Taipei-1', 3, 28, 18, 57),
 ('SLO', 'Slovenia', 8, 13, 27, 58),
 ('SVK', 'Slovakia', 15, 19, 13, 59),
 ('AZE', 'Azerbaijan', 7, 12, 25, 60),
 ('POR', 'Portugal', 4, 11, 26, 61),
 ('INA', 'Indonesia', 11, 17, 13, 61),
 ('BAH', 'Bahamas', 14, 11, 15, 62),
 ('LAT', 'Latvia', 3, 19, 13, 63),
 ('IRL', 'Ireland', 9, 13, 13, 63),
 ('UZB', 'Uzbekistan', 10, 7, 17, 64),
 ('TTO', 'Trinidad and Tobago', 7, 8, 17, 65),
 ('GEO', 'Georgia', 8, 6, 18, 65),
 ('CHI', 'Chile', 3, 9, 20, 65),
 ('THA', 'Thailand', 9, 8, 13, 66),
 ('ANZ', 'Australasia', 20, 4, 5, 67),
 ('COL', 'Colombia', 5, 9, 14, 68),
 ('EGY', 'Egypt', 7, 8, 12, 69),
 ('MGL', 'Mongolia', 2, 10, 14, 70),
 ('MAR', 'Morocco', 6, 5, 12, 71),
 ('GHA', 'Ghana', 0, 1, 22, 71),
 ('ZIM', 'Zimbabwe', 17, 4, 1, 72),
 ('CMR', 'Cameroon', 20, 1, 1, 72),
 ('PAR', 'Paraguay', 0, 17, 0, 73),
 ('ISL', 'Iceland', 0, 15, 2, 73),
 ('ALG', 'Algeria', 5, 4, 8, 73),
 ('MAS', 'Malaysia', 0, 11, 5, 74),
 ('ARM', 'Armenia', 2, 5, 9, 74),
 ('VEN', 'Venezuela', 2, 3, 10, 75),
 ('PER', 'Peru', 1, 14, 0, 75),
 ('MNE', 'Montenegro', 0, 14, 0, 76),
 ('TUN', 'Tunisia', 3, 3, 7, 77),
 ('FIJ', 'Fiji', 13, 0, 0, 77),
 ('BOH', 'Bohemia', 0, 1, 11, 78),
 ('PHI', 'Philippines', 0, 3, 7, 79),
 ('SGP', 'Singapore', 1, 4, 4, 80),
 ('PUR', 'Puerto Rico', 1, 2, 6, 80),
 ('LIE', 'Liechtenstein', 2, 2, 5, 80),
 ('ISR', 'Israel', 1, 1, 7, 80),
 ('MDA', 'Moldova', 0, 3, 5, 81),
 ('LUX', 'Luxembourg', 4, 4, 0, 81),
 ('UGA', 'Uganda', 2, 3, 2, 82),
 ('HAI', 'Haiti', 1, 1, 5, 82),
 ('DOM', 'Dominican Republic', 3, 2, 2, 82),
 ('KSA', 'Saudi Arabia', 0, 1, 5, 83),
 ('WIF', 'Circus', 0, 0, 5, 84),
 ('QAT', 'Qatar', 0, 1, 4, 84),
 ('IOA', 'Individual Olympic Athletes', 1, 1, 3, 84),
 ('VIE', 'Vietnam', 1, 3, 0, 85),
 ('TJK', 'Taji

## 2.15) Let's list down total gold, silver and bronze medals won by each country corresponding to each olympic games

In [31]:
query_15 = """
SELECT NOC, Team, Games,
  SUM(CASE WHEN Medal = 'Gold' THEN 1 ELSE 0 END) AS NumGoldMedals,
  SUM(CASE WHEN Medal = 'Silver' THEN 1 ELSE 0 END) AS NumSilverMedals,
  SUM(CASE WHEN Medal = 'Bronze' THEN 1 ELSE 0 END) AS NumBronzeMedals,
  DENSE_RANK() OVER (ORDER BY SUM(CASE WHEN Medal IN ('Gold', 'Silver', 'Bronze') THEN 1 ELSE 0 END) DESC) AS Ranking
FROM athlete_events
WHERE Games = '1896 Summer'
GROUP BY NOC
"""

%sql $query_15

 * sqlite:///athlete_events.db
Done.


[('GRE', 'Greece', '1896 Summer', 10, 18, 20, 1),
 ('GER', 'Germany', '1896 Summer', 25, 5, 2, 2),
 ('USA', 'United States', '1896 Summer', 11, 7, 2, 3),
 ('FRA', 'France', '1896 Summer', 5, 4, 2, 4),
 ('GBR', 'Great Britain', '1896 Summer', 3, 3, 3, 5),
 ('HUN', 'Hungary', '1896 Summer', 2, 1, 3, 6),
 ('DEN', 'Denmark', '1896 Summer', 1, 2, 3, 6),
 ('AUT', 'Austria', '1896 Summer', 2, 1, 2, 7),
 ('SUI', 'Switzerland', '1896 Summer', 1, 2, 0, 8),
 ('AUS', 'Australia', '1896 Summer', 2, 0, 1, 8),
 ('SWE', 'Sweden', '1896 Summer', 0, 0, 0, 9),
 ('ITA', 'Italy', '1896 Summer', 0, 0, 0, 9)]

## It costs too much, computationally speaking, to retrieve all the games, so we selected only one of them. But the same logic applies if we want to display them all, it is enough to remove the 'WHERE' clause

## 2.16) let's identify which country won the most gold, most silver and most bronze medals in each olympic games

In [37]:
query_16 = """
SELECT Games, CountryGoldMedals, CountrySilverMedals, CountryBronzeMedals
FROM (
  SELECT Games,
    (SELECT Team FROM (
      SELECT Team, COUNT(*) AS NumMedals
      FROM athlete_events
      WHERE Medal = 'Gold' AND Games = medal_counts.Games
      GROUP BY Team
    ) AS gold_medals
    WHERE gold_medals.NumMedals = (SELECT MAX(NumMedals) FROM (
      SELECT COUNT(*) AS NumMedals
      FROM athlete_events
      WHERE Medal = 'Gold' AND Games = medal_counts.Games
      GROUP BY Team
    ) AS gold_medals)) AS CountryGoldMedals,
    (SELECT Team FROM (
      SELECT Team, COUNT(*) AS NumMedals
      FROM athlete_events
      WHERE Medal = 'Silver' AND Games = medal_counts.Games
      GROUP BY Team
    ) AS silver_medals
    WHERE silver_medals.NumMedals = (SELECT MAX(NumMedals) FROM (
      SELECT COUNT(*) AS NumMedals
      FROM athlete_events
      WHERE Medal = 'Silver' AND Games = medal_counts.Games
      GROUP BY Team
    ) AS silver_medals)) AS CountrySilverMedals,
    (SELECT Team FROM (
      SELECT Team, COUNT(*) AS NumMedals
      FROM athlete_events
      WHERE Medal = 'Bronze' AND Games = medal_counts.Games
      GROUP BY Team
    ) AS bronze_medals
    WHERE bronze_medals.NumMedals = (SELECT MAX(NumMedals) FROM (
      SELECT COUNT(*) AS NumMedals
      FROM athlete_events
      WHERE Medal = 'Bronze' AND Games = medal_counts.Games
      GROUP BY Team
    ) AS bronze_medals)) AS CountryBronzeMedals
  FROM (
    SELECT Games, Team, Medal, COUNT(*) AS NumMedals
    FROM athlete_events
    GROUP BY Games, Team, Medal
  ) AS medal_counts
  GROUP BY Games
)
"""

%sql $query_16

 * sqlite:///athlete_events.db
Done.


[('1896 Summer', 'Germany', 'Greece', 'Greece'),
 ('1900 Summer', 'France', 'France', 'France'),
 ('1904 Summer', 'United States', 'United States', 'United States'),
 ('1906 Summer', 'France', 'Denmark', 'France'),
 ('1908 Summer', 'Great Britain', 'Great Britain', 'Great Britain'),
 ('1912 Summer', 'Sweden', 'Denmark', 'Great Britain'),
 ('1920 Summer', 'United States', 'France', 'Sweden'),
 ('1924 Summer', 'United States', 'France', 'United States'),
 ('1924 Winter', 'Great Britain', 'United States', 'Great Britain'),
 ('1928 Summer', 'United States', 'Great Britain', 'Germany'),
 ('1928 Winter', 'Canada', 'Sweden', 'Switzerland'),
 ('1932 Summer', 'United States', 'United States', 'United States'),
 ('1932 Winter', 'Canada', 'United States', 'Germany'),
 ('1936 Summer', 'Germany', 'Germany', 'Germany'),
 ('1936 Winter', 'Great Britain', 'Canada', 'United States'),
 ('1948 Summer', 'United States', 'Great Britain', 'United States'),
 ('1948 Winter', 'Canada', 'Czechoslovakia', 'Switzerland'),
 ('1952 Summer', 'United States', 'Soviet Union', 'Hungary'),
 ('1952 Winter', 'Canada', 'United States', 'Sweden'),
 ('1956 Summer', 'Soviet Union', 'Soviet Union', 'Soviet Union'),
 ('1956 Winter', 'Soviet Union', 'United States', 'Canada'),
 ('1960 Summer', 'United States', 'Soviet Union', 'Soviet Union'),
 ('1960 Winter', 'United States', 'Canada', 'Soviet Union'),
 ('1964 Summer', 'United States', 'Soviet Union', 'Soviet Union'),
 ('1964 Winter', 'Soviet Union', 'Sweden', 'Czechoslovakia'),
 ('1968 Summer', 'United States', 'Soviet Union', 'Soviet Union'),
 ('1968 Winter', 'Soviet Union', 'Czechoslovakia', 'Canada'),
 ('1972 Summer', 'Soviet Union', 'East Germany', 'Soviet Union'),
 ('1972 Winter', 'Soviet Union', 'United States', 'Czechoslovakia'),
 ('1976 Summer', 'Soviet Union', 'Soviet Union', 'Soviet Union'),
 ('1976 Winter', 'Soviet Union', 'Czechoslovakia', 'West Germany'),
 ('1980 Summer', 'Soviet Union', 'Soviet Union', 'Soviet Union'),
 ('1980 Winter', 'United States', 'Soviet Union', 'Sweden'),
 ('1984 Summer', 'United States', 'United States', 'West Germany'),
 ('1984 Winter', 'Soviet Union', 'Czechoslovakia', 'Sweden'),
 ('1988 Summer', 'Soviet Union', 'Soviet Union', 'Soviet Union'),
 ('1988 Winter', 'Soviet Union', 'Finland', 'Sweden'),
 ('1992 Summer', 'Unified Team', 'Unified Team', 'United States'),
 ('1992 Winter', 'Unified Team', 'Canada', 'Czechoslovakia'),
 ('1994 Winter', 'Sweden', 'Canada', 'Finland'),
 ('1996 Summer', 'United States', 'China', 'Australia'),
 ('1998 Winter', 'United States', 'Canada', 'Finland'),
 ('2000 Summer', 'United States', 'Australia', 'Germany'),
 ('2002 Winter', 'Canada', 'United States', 'Russia'),
 ('2004 Summer', 'United States', 'Australia', 'Russia'),
 ('2006 Winter', 'Sweden', 'Finland', 'United States'),
 ('2008 Summer', 'United States', 'United States', 'United States'),
 ('2010 Winter', 'Canada', 'United States', 'Finland'),
 ('2012 Summer', 'United States', 'United States', 'Australia'),
 ('2014 Winter', 'Canada', 'Sweden', 'Finland'),
 ('2016 Summer', 'United States', 'France', 'Germany')]

## 2.17) Let's now fetch which countries have never won gold medal but have won silver/bronze medals?

In [44]:
query_17 = """
SELECT NOC, Team,
  SUM(CASE WHEN Medal = 'Gold' THEN 1 ELSE 0 END) AS NumGoldMedals,
  SUM(CASE WHEN Medal = 'Silver' THEN 1 ELSE 0 END) AS NumSilverMedals,
  SUM(CASE WHEN Medal = 'Bronze' THEN 1 ELSE 0 END) AS NumBronzeMedals,
  DENSE_RANK() OVER (ORDER BY COUNT(*) DESC) AS Ranking
FROM athlete_events
WHERE NOC NOT IN (
  SELECT NOC
  FROM athlete_events
  WHERE Medal = 'Gold'
) AND Medal IN ('Silver', 'Bronze')
GROUP BY NOC
"""

%sql $query_17

 * sqlite:///athlete_events.db
Done.


[('GHA', 'Ghana', 0, 1, 22, 1),
 ('PAR', 'Paraguay', 0, 17, 0, 2),
 ('ISL', 'Iceland', 0, 15, 2, 2),
 ('MAS', 'Malaysia', 0, 11, 5, 3),
 ('MNE', 'Montenegro', 0, 14, 0, 4),
 ('BOH', 'Bohemia', 0, 1, 11, 5),
 ('PHI', 'Philippines', 0, 3, 7, 6),
 ('MDA', 'Moldova', 0, 3, 5, 7),
 ('KSA', 'Saudi Arabia', 0, 1, 5, 8),
 ('WIF', 'West Indies Federation', 0, 0, 5, 9),
 ('QAT', 'Qatar', 0, 1, 4, 9),
 ('NAM', 'Namibia', 0, 4, 0, 10),
 ('LIB', 'Lebanon', 0, 2, 2, 10),
 ('KGZ', 'Kyrgyzstan', 0, 1, 2, 11),
 ('ZAM', 'Zambia', 0, 1, 1, 12),
 ('UAR', 'United Arab Republic', 0, 1, 1, 12),
 ('TAN', 'Tanzania', 0, 2, 0, 12),
 ('SRI', 'Sri Lanka', 0, 2, 0, 12),
 ('NIG', 'Niger', 0, 1, 1, 12),
 ('KUW', 'Kuwait', 0, 0, 2, 12),
 ('AFG', 'Afghanistan', 0, 0, 2, 12),
 ('TOG', 'Togo', 0, 0, 1, 13),
 ('TGA', 'Tonga', 0, 1, 0, 13),
 ('SUD', 'Sudan', 0, 1, 0, 13),
 ('SEN', 'Senegal', 0, 1, 0, 13),
 ('MRI', 'Mauritius', 0, 0, 1, 13),
 ('MON', 'Monaco', 0, 0, 1, 13),
 ('MKD', 'Macedonia', 0, 0, 1, 13),
 ('ISV', 'United States Virgin Islands', 0, 1, 0, 13),
 ('IRQ', 'Iraq', 0, 0, 1, 13),
 ('GUY', 'Guyana', 0, 0, 1, 13),
 ('GUA', 'Guatemala', 0, 1, 0, 13),
 ('GAB', 'Gabon', 0, 1, 0, 13),
 ('ERI', 'Eritrea', 0, 0, 1, 13),
 ('DJI', 'Djibouti', 0, 0, 1, 13),
 ('CYP', 'Cyprus', 0, 1, 0, 13),
 ('BOT', 'Botswana', 0, 1, 0, 13),
 ('BER', 'Bermuda', 0, 0, 1, 13),
 ('BAR', 'Barbados', 0, 0, 1, 13),
 ('AHO', 'Netherlands Antilles', 0, 1, 0, 13)]

## 2.18) In which sport India has won the most medals ?

In [51]:
query_18 = """
SELECT Sport, COUNT(*) AS NumMedals
FROM athlete_events
WHERE Team = 'India'
GROUP BY Sport
ORDER BY NumMedals DESC
LIMIT 1
"""

%sql $query_18

 * sqlite:///athlete_events.db
Done.


[('Hockey', 345)]

## 2.19) Let's break down all olympic games where India won medal for Hockey and how many medals in each olympic games

In [60]:
query_19 = """
SELECT Team, Games, Sport, COUNT(*) AS NumMedals,
  DENSE_RANK() OVER (ORDER BY COUNT(*) DESC) AS Ranking
FROM athlete_events
WHERE Team = 'India' AND Sport = 'Hockey'
GROUP BY Team, Games, Sport
"""

%sql $query_19

 * sqlite:///athlete_events.db
Done.


[('India', '2016 Summer', 'Hockey', 32, 1),
 ('India', '1980 Summer', 'Hockey', 30, 2),
 ('India', '1948 Summer', 'Hockey', 20, 3),
 ('India', '1936 Summer', 'Hockey', 19, 4),
 ('India', '1956 Summer', 'Hockey', 17, 5),
 ('India', '1968 Summer', 'Hockey', 16, 6),
 ('India', '1976 Summer', 'Hockey', 16, 6),
 ('India', '1984 Summer', 'Hockey', 16, 6),
 ('India', '1988 Summer', 'Hockey', 16, 6),
 ('India', '1996 Summer', 'Hockey', 16, 6),
 ('India', '2004 Summer', 'Hockey', 16, 6),
 ('India', '2012 Summer', 'Hockey', 16, 6),
 ('India', '1932 Summer', 'Hockey', 15, 7),
 ('India', '1964 Summer', 'Hockey', 15, 7),
 ('India', '1992 Summer', 'Hockey', 15, 7),
 ('India', '2000 Summer', 'Hockey', 15, 7),
 ('India', '1928 Summer', 'Hockey', 14, 8),
 ('India', '1952 Summer', 'Hockey', 14, 8),
 ('India', '1972 Summer', 'Hockey', 14, 8),
 ('India', '1960 Summer', 'Hockey', 13, 9)]